src/treelearner/monotone_constraints.hpp

```diff
@@ -55,14 +92,368 @@ class LeafConstraints {
+  const ConstraintEntry& Get(int leaf_idx) const override { return entries_[leaf_idx]; }
 
- private:
+ protected:
   int num_leaves_;
   std::vector<ConstraintEntry> entries_;
 };
 
+class FastLeafConstraints : public BasicLeafConstraints {
+ public:
+  explicit FastLeafConstraints(const Config* config, int num_leaves)
+      : BasicLeafConstraints(num_leaves), config_(config) {
+    leaf_is_in_monotone_subtree_.resize(num_leaves_, false);
+    node_parent_.resize(num_leaves_ - 1, -1);
+    leaves_to_update_.reserve(num_leaves_);
+  }
+
+  void Reset() override {
+    BasicLeafConstraints::Reset();
+    std::fill_n(leaf_is_in_monotone_subtree_.begin(), num_leaves_, false);
+    std::fill_n(node_parent_.begin(), num_leaves_ - 1, -1);
+    leaves_to_update_.clear();
+  }
+
+  void BeforeSplit(const Tree* tree, int leaf, int new_leaf,
+                   int8_t monotone_type) override {
+    if (monotone_type != 0 || leaf_is_in_monotone_subtree_[leaf]) {
+      leaf_is_in_monotone_subtree_[leaf] = true;
+      leaf_is_in_monotone_subtree_[new_leaf] = true;
+    }
+#ifdef DEBUG
+    CHECK(new_leaf - 1 >= 0);
+    CHECK((unsigned int)(new_leaf - 1) < node_parent_.size());
+#endif
+    node_parent_[new_leaf - 1] = tree->leaf_parent(leaf);
+  }
+
+  void UpdateConstraintsWithOutputs(bool is_numerical_split, int leaf,
+                                    int new_leaf, int8_t monotone_type,
+                                    double right_output, double left_output) {
+    entries_[new_leaf] = entries_[leaf];
+    if (is_numerical_split) {
+      if (monotone_type < 0) {
+        entries_[leaf].UpdateMin(right_output);
+        entries_[new_leaf].UpdateMax(left_output);
+      } else if (monotone_type > 0) {
+        entries_[leaf].UpdateMax(right_output);
+        entries_[new_leaf].UpdateMin(left_output);
+      }
+    }
+  }
+
+  std::vector<int> Update(const Tree* tree, bool is_numerical_split, int leaf,
+                          int new_leaf, int8_t monotone_type,
+                          double right_output, double left_output,
+                          int split_feature, const SplitInfo& split_info,
+                          const std::vector<SplitInfo>& best_split_per_leaf) override{
+    leaves_to_update_.clear();
+    if (leaf_is_in_monotone_subtree_[leaf]) {
+      UpdateConstraintsWithOutputs(is_numerical_split, leaf, new_leaf,
+                                   monotone_type, right_output, left_output);
+
+      GoUpToFindLeavesToUpdate(tree, tree->leaf_parent(new_leaf), split_feature,
+                               split_info, split_info.threshold,
+                               best_split_per_leaf);
+    }
+    return leaves_to_update_;
+  }
+
+  bool OppositeChildShouldBeUpdated(
+      bool is_split_numerical,
+      const std::vector<int>& features_of_splits_going_up_from_original_leaf,
+      int inner_feature,
+      std::vector<bool>& was_original_leaf_right_child_of_split,
```

Comment:
use `std::vector<T>*` if you want to change its content inside the function. 
you can use `std::vector<T>& ref_a = *a` in the function, to access the vector.

Reply:
if you don't want to change its content, please add `const`.